neqo-http3/src/recv_message.rs

```diff
@@ -337,115 +345,93 @@ impl RecvMessage {
+
+        if *state & bit == 0 || !pseudo {
+            *state |= bit;
+            Ok(pseudo)
+        } else {
+            Err(Error::HttpGeneralProtocolStream)
         }
-        let mut seen_non_colon_header = false;
-        let excluded_headers = vec![
-            "connection",
-            "host",
-            "keep-alive",
-            "proxy-connection",
-            "te",
-            "transfer-encoding",
-            "upgrade",
-            "accept-encoding",
-        ];
-        let excluded_header_byte = vec![0x0, 0x10, 0x13, 0x3a];
+    }
+
+    fn headers_valid(&self, headers: &[Header]) -> Res<bool> {
         let mut method_value: Option<&String> = None;
+        let mut pseudo_state = 0;
         for header in headers {
-            let mut header_byte_iterator = header.0.bytes();
-            let is_colon_header = header.0.starts_with(':');
-            if is_colon_header {
-                if seen_non_colon_header {
-                    qdebug!([self], "headers_valid - pseudo header fields after fields");
-                    return Err(Error::HttpGeneralProtocolStream);
-                }
+            let is_pseudo = self.track_pseudo(&header.0, &mut pseudo_state)?;
 
-                if let hash_map::Entry::Occupied(mut e) = pseudo_header_map.entry(&header.0) {
-                    if *e.get_mut() {
-                        return Err(Error::HttpGeneralProtocolStream);
-                    }
-                    *e.get_mut() = true;
-                } else {
+            let mut bytes = header.0.bytes();
+            if is_pseudo {
+                if (pseudo_state & (PSEUDO_HEADER_UNDEFINED | REGULAR_HEADER)) != 0 {
                     return Err(Error::HttpGeneralProtocolStream);
                 }
 
                 if header.0 == ":method" {
-                    method_value = Some(&header.1)
+                    method_value = Some(&header.1);
                 }
-
-                // skip the first ":"
-                header_byte_iterator.next();
-            } else {
-                seen_non_colon_header = true;
+                let _ = bytes.next();
             }
 
-            for byte in header_byte_iterator {
-                if byte.is_ascii_uppercase() {
-                    qdebug!(
-                        [self],
-                        "headers_valid - found uppercase in header:'{}'",
-                        header.0
-                    );
-                    return Err(Error::HttpGeneralProtocolStream);
-                }
-
-                if excluded_header_byte
-                    .iter()
-                    .any(|&excluded| excluded == byte)
-                {
-                    qdebug!([self], "headers_valid - found an excluded byte");
-                    return Err(Error::HttpGeneralProtocolStream);
-                }
+            if bytes.any(|b| matches!(b, 0 | 0x10 | 0x13 | 0x3a | 0x41..=0x5a)) {
+                return Err(Error::HttpGeneralProtocolStream); // illegal characters.
```

Comment:
I think that we need a better error for this header validation.  It can be treated as a stream error and the same error code, but it would be good to have better precision regarding local errors.  If we had `Error::InvalidHeaders` or similar we could at least trace errors back to this code.

Reply:
I'll add a new error code `HttpInvalidHeader`.